### PR TITLE
Add parse tree function

### PR DESCRIPTION
**Pull request by oleggromov**
I once had the following problem as an interview question.

```
First
  Second
    Third
  Second #2 
First
```

Goal is to parse this string into a tree. There's `2 × N` spaces in the beginning of each line, where `N` is a multiple of 2.

## Interview
At the interview time, I couldn't solve it. I think I got lost in maintaining state and also tedious string manipulations like counting spaces, matching regexp, etc. This time, it also took me surprisingly much longer than I expected.

## Solution
The first and the one I actually implemented is the following algorithm:

1. Consume the string in chunks separated with `\n`
2. For each chunk, detect it's `depth` by counting space characters
3. Keep track of `path` that represents the deepest level we got to
4. Create a `TreeNode` from each chunk
5. Add it as a child to the appropriate leaf
6. Modify `path` if required:
  6.1 Trim it next chunk is on the lower depth level
  6.2 Make it deeper by 1 if the next chunk is a child of the latest one

I think steps 5) and 6) took me the longest.

### Visualizing state
This was the most tricky thing. I had a correct idea in mind for long, but couldn't realize I need to add each chunk to the leaf that's second to last and push that node to be the last leaf in the path.

IMG_8456

## Questions
I would be grateful if you could:
1. Tell me why this was hard to implement in the first place? Maybe I am missing some fundamental pieces of knowledge?
2. What data structure would you use to represent the "path"? 
3. What would you change?

### CODE REVIEW

File: parse-tree.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+const parseTree = str => {
+  const root = new TreeNode
+  const path = [root]
+
+  for (let chunk of str.split('\n')) {
+    const match = chunk.match(/(\s*)(.*)/)
+
+    if (match && match[1].length % 2 !== 0) {
+      throw new Error(`Illegal indentation: '${chunk}'`)
+    }
```
Comment (oleggromov) [inline parse-tree.js]: This is just for the sake of addressing corrupt input.

**Comment by vitkarpov** [comment]
> Tell me why this was hard to implement in the first place?

Actually, I struggled with it at first too. What I noticed personally, I wanted to use recursion since it looks like that but actually it's not, at least the current solution is just a linear scan of lines (with a stack to keep track of levels).

**Comment by oleggromov** [comment]
@vitkarpov I like your solution. It's more concise and clever than mine, mostly because you managed to avoid inside-the-loop conditionals and exceptions. 

> I use kind of "a stack" to keep track of levels explicitly and you're just cutting the array
In fact, I started thinking of stack exactly, but later understood that I don't want to perform these few `pop` operations and started just slicing like I do now. Then I realized it's not how stacks are supposed to work and renamed it to path. But yes, it's absolutely similar in functionality to what you do.

>  I wanted to use recursion since it looks like that
I struggled because of more trivial levels. Either I couldn't fit the entire algorithm in mind, or didn't properly model it beforehand... I don't know.

**Comment by oleggromov** [comment]
Maybe last but not least, we tend to ignore computational complexity of things like `str.split('\n')` (what's that, in fact, I guess `O(str.length)`?) and `str.match(/(\s*)(.*)/` (this must be take longer than linear time to execute, I would assume). What's the right way of estimating complexity in such cases?

**Comment by vitkarpov** [comment]
> What's the right way of estimating complexity in such cases?

Well, it's a good question 😃 

Regards to splitting it should be `O(str.length ^ 2)` since we need to not only scan the entire string but to create `O(str.length)` (one-char strings in the worst case) strings as well.

It's not clear about regexps though, so I'd ask interview is it okay to use such a regexp here or extract a substring manually (in that case the complexity should be more or less clear).

**Comment by oleggromov** [comment]
@vitkarpov, good ideas, thanks. I'll close this for a while and maybe will get back to this task later when I forget it and will have a chance to do essentially from scratch :)

**Pull request merged**